### DISCUSSION TITLE
idea: extract individual contract read errors from `useContractReads`

### DISCUSSION THREAD

**Discussion by 0xVesper** [wevm/wagmi#1159] [Ideas]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Package Version

0.6.7

### Current Behavior

Hello,

I've tried adding an `onError` to each read of `useContractReads`, but I don't believe the feature exists.

Is there a way to retrieve the error from each read, instead of aborting and only getting the first error returned?

Thank you 🤝

### Expected Behavior

`useContractreads` only will use `onError` if `allowFailure: false` and aborts all of the reads upon a single error. I need the reads to happen for each call passed in, and return an error for each call that happens if it exists.

### Steps To Reproduce

For example:

```
const { data, isError, isLoading } = useContractReads({
    contracts: [
      {
        ...wagmigotchiContract,
        functionName: 'getAlive',
        onError(error) {       // this does nothing since feature doesn't exist
               console.log("Error thrown: ", error);
        },
      },
    ],
  })
```

It currently forces use of `allowFailure`, will stop calling the reads after the first error is thrown, and only show said first error: 

```
const { data, isError, isLoading } = useContractReads({
    contracts: [
      {
        ...wagmigotchiContract,
        functionName: 'getAlive'
      },
    allowFailure: false,
    onError(error) {     // only works on first read error and stops all reads
      console.log("Error thrown: ", error);
    },
    ],
  })
```

**Comment by jxom** [comment]
Thanks for this issue!

Definitely agree it would be nice to have a way to extract the error of a contract read when `allowFailure` is truthy.

The way `onError` behaves right now is that it is invoked when the hook's function throws an error, but in this case, it will not as `allowFailure` is truthy. Perhaps we should have another option to handle this, like `onReadError` or something.

I'd say this issue is more of a feature request than a bug, so I am going to move this to a discussion under "Ideas" & I'll add it to the wagmi roadmap.

**Comment by 0xVesper** [comment]
The returned error per read shows in the console when `allowFailure: true`. How can we catch this and use it?

**Reply by jxom** [reply]
You can't yet, but we can definitely make this happen.

**Reply by 0xVesper** [reply]
I see... it's a necessity in my case since a read is happening per front-end component, so if I don't catch the error I can't display the revert reason for UX. 

How could I go about contributing to this feature? Or even a global feature for showing a toast or something whenever any revert happens, inside whatever is handling `console.log` of the revert..

**Comment by iainnash** [comment]
Can the return value be the object or an javascript `Error` object when in this mode?

That way you can iterate the response and handle the Error objects as needed. Multicall3 supports this in the return value.

We recently ran into this with a front-end.

**Reply by jxom** [reply]
On the roadmap for v1!

**Comment by j5h46bd** [comment]
Why is allowFailure on reads but not read (singular)? Is there a more appropriate way to suppress a particular common contract reversion?
